### PR TITLE
prov/efa: Add unit tests for efa_msg

### PR DESCRIPTION
**Pull request by jiaxiyan**
Add unit tests for efa_msg. Use FI_EP_RDM as ep type and override efa_msg_ops.

### CODE REVIEW

File: prov/efa/test/efa_unit_test_msg.c [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+	desc = fi_mr_desc(send_buff.mr);
+	efa_unit_test_construct_msg(&msg, &iov, 1, addr, NULL, 0, &desc);
+
+	ret = fi_recvmsg(resource->ep, &msg, 0);
+	assert_int_equal(ret, 0);
+
+	efa_unit_test_buff_destruct(&send_buff);
+}
+
+static void test_efa_msg_send_impl(struct efa_resource *resource,
```
Comment (shijin-aws) [inline prov/efa/test/efa_unit_test_msg.c:99]: This doesn't look a `send_impl`, it looks `send_prep`

File: prov/efa/test/efa_unit_test_msg.c [hunk 2]
```diff
@@ -0,0 +1,225 @@
+/* SPDX-License-Identifier: BSD-2-Clause OR GPL-2.0-only */
+/* SPDX-FileCopyrightText: Copyright Amazon.com, Inc. or its affiliates. All
+ * rights reserved. */
+
+#include "efa_unit_tests.h"
+#include "ofi_util.h"
+
+static void test_efa_msg_recv_impl(struct efa_resource *resource,
```
Comment (shijin-aws) [inline prov/efa/test/efa_unit_test_msg.c:8]: same comment as the `send_impl`

File: prov/efa/test/efa_unit_tests.c [hunk 3]
```diff
@@ -208,6 +208,13 @@ int main(void)
...
 		cmocka_unit_test_setup_teardown(test_efa_rdm_peer_append_overflow_pke_to_recvwin, efa_unit_test_mocks_setup, efa_unit_test_mocks_teardown),
 		cmocka_unit_test_setup_teardown(test_efa_rdm_pke_handle_longcts_rtm_send_completion, efa_unit_test_mocks_setup, efa_unit_test_mocks_teardown),
+		cmocka_unit_test_setup_teardown(test_efa_msg_fi_recv, efa_unit_test_mocks_setup, efa_unit_test_mocks_teardown),
+		cmocka_unit_test_setup_teardown(test_efa_msg_fi_recvv, efa_unit_test_mocks_setup, efa_unit_test_mocks_teardown),
+		cmocka_unit_test_setup_teardown(test_efa_msg_fi_recvmsg, efa_unit_test_mocks_setup, efa_unit_test_mocks_teardown),
+		cmocka_unit_test_setup_teardown(test_efa_msg_fi_send, efa_unit_test_mocks_setup, efa_unit_test_mocks_teardown),
+		cmocka_unit_test_setup_teardown(test_efa_msg_fi_sendv, efa_unit_test_mocks_setup, efa_unit_test_mocks_teardown),
+		cmocka_unit_test_setup_teardown(test_efa_msg_fi_sendmsg, efa_unit_test_mocks_setup, efa_unit_test_mocks_teardown),
+		cmocka_unit_test_setup_teardown(test_efa_msg_fi_senddata, efa_unit_test_mocks_setup, efa_unit_test_mocks_teardown),
```
Comment (shijin-aws) [inline prov/efa/test/efa_unit_tests.c:217]: Do you want to add tests for inject* calls?

Comment (jiaxiyan) [reply prov/efa/test/efa_unit_tests.c:217]: I am not sure how to test the inject as I am using FI_EP_DGRAM as the ep type right now.

Comment (shijin-aws) [reply prov/efa/test/efa_unit_tests.c:217]: easy thing: use FI_EP_RDM, but override your ep function table before you making the send/recv call.

```
// get ep from your ep_fid
(*ep)->msg = &efa_msg_ops;
(*ep)->rma = &efa_rma_ops;
```

Comment (shijin-aws) [reply prov/efa/test/efa_unit_tests.c:217]: We should finally make DGRAM support inject after onboarding FI_CONTEXT2, you can do this override before that happens.

Comment (shijin-aws) [reply prov/efa/test/efa_unit_tests.c:217]: You need to make `efa_msg_ops` in a header for your unit test to include

**Review by shijin-aws — APPROVED** [review]
Thanks!

**Comment by shijin-aws** [comment]
Ideally we should add some more validation in the rdma-core level to make sure sge, imm data are set correctly, but that is likely beyond the scope of this PR. We will start from this one and explore that further validation as follow-ups.

**Pull request merged**